api/utils/keys/yubikey.go

```diff
@@ -0,0 +1,384 @@
+		return nil, trace.NotFound("YubiKey certificate slot contained unknown certificate:\n%+v", cert)
+	}
+
+	// Attest the key to make sure it hasn't been imported.
+	slotCert, err := yk.Attest(slot)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	attestationCert, err := yk.AttestationCertificate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if _, err = piv.Verify(attestationCert, slotCert); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Verify that the slot's certs have the same public key, otherwise the key
+	// may have been generated by a non-teleport client.
+	if pubComparer, ok := cert.PublicKey.(interface{ Equal(x crypto.PublicKey) bool }); !ok {
+		return nil, trace.BadParameter("certificate's public key of type %T is not a supported public key", cert.PublicKey)
+	} else if !pubComparer.Equal(slotCert.PublicKey) {
+		return nil, trace.NotFound("YubiKey slot contains mismatched certificates and must be regenerated")
+	}
+
+	return newYubiKeyPrivateKey(y, slot, slotCert.PublicKey)
+}
+
+// open a connection to yubiKey PIV module. The returned connection should be closed once
+// it's been used. The yubiKey PIV module itself takes some additional time to handle closed
+// connections, so we use a retry loop to give the PIV module time to close prior connections.
+func (y *yubiKey) open() (yk *piv.YubiKey, err error) {
+	isRetryError := func(err error) bool {
+		retryError := "connecting to smart card: the smart card cannot be accessed because of other connections outstanding"
+		return strings.Contains(err.Error(), retryError)
+	}
+
+	var maxRetries int = 100
+	for i := 0; i < maxRetries; i++ {
+		yk, err = piv.Open(y.card)
+		if err == nil {
+			return yk, nil
+		}
+
+		if !isRetryError(err) {
+			return nil, trace.Wrap(err)
+		}
+
+		time.Sleep(time.Millisecond * 100)
+	}
+
+	return nil, trace.Wrap(err)
+}
+
+// findYubiKey finds a yubiKey PIV card by serial number. If no serial
+// number is provided, the first yubiKey found will be returned.
+func findYubiKey(serialNumber uint32) (*yubiKey, error) {
+	yubiKeyCards, err := findYubiKeyCards()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	for _, card := range yubiKeyCards {
+		y, err := newYubiKey(card)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if serialNumber == 0 || y.serialNumber == serialNumber {
+			return y, nil
+		}
+	}
+
+	return nil, trace.NotFound("no yubiKey device found with serial number %q", serialNumber)
+}
+
+// findYubiKeyCards returns a list of connected yubiKey PIV card names.
+func findYubiKeyCards() ([]string, error) {
+	cards, err := piv.Cards()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var yubiKeyCards []string
+	for _, card := range cards {
+		if strings.Contains(strings.ToLower(card), PIVCardTypeYubiKey) {
+			yubiKeyCards = append(yubiKeyCards, card)
+		}
+	}
+
+	if len(yubiKeyCards) == 0 {
+		return nil, trace.NotFound("no yubiKey devices found")
+	}
+
```

Comment:
"No cards connected" can be signaled by returning zero cards - otherwise returning `[]string{}, nil` would be possible but invalid.

```suggestion
```